crates/allocator/src/bump.rs

```diff
@@ -0,0 +1,241 @@
+//!
+//! Its goal to have a much smaller footprint than the admittedly more full-featured `wee_alloc`
+//! allocator which is currently being used by ink! smart contracts.
+//!
+//! The heap which is used by this allocator is built from pages of Wasm memory (each page is `64KiB`).
+//! We will request new pages of memory as needed until we run out of memory, at which point we
+//! will crash with an `OOM` error instead of freeing any memory.
+
+use core::alloc::{
+    GlobalAlloc,
+    Layout,
+};
+
+/// A page in Wasm is `64KiB`
+const PAGE_SIZE: usize = 64 * 1024;
+
+static mut INNER: InnerAlloc = InnerAlloc::new();
+
+/// A bump allocator suitable for use in a Wasm environment.
+pub struct BumpAllocator;
+
+unsafe impl GlobalAlloc for BumpAllocator {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        match INNER.alloc(layout) {
+            Some(start) => start as *mut u8,
+            None => core::ptr::null_mut(),
+        }
+    }
+
+    #[inline]
+    unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
+        // A new page in Wasm is guaranteed to already be zero initialized, so we can just use our
+        // regular `alloc` call here and save a bit of work.
+        //
+        // See: https://webassembly.github.io/spec/core/exec/modules.html#growing-memories
+        self.alloc(layout)
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+}
+
+#[cfg_attr(feature = "std", derive(Debug, Copy, Clone))]
+struct InnerAlloc {
+    /// Points to the start of the next available allocation.
+    next: usize,
+
+    /// The address of the upper limit of our heap.
+    upper_limit: usize,
+}
+
+impl InnerAlloc {
+    const fn new() -> Self {
+        Self {
+            next: 0,
+            upper_limit: 0,
+        }
+    }
+
+    cfg_if::cfg_if! {
+        if #[cfg(all(not(feature = "std"), target_arch = "wasm32"))] {
+            /// Request a `pages` number of pages of Wasm memory. Each page is `64KiB` in size.
+            ///
+            /// Returns `None` if a page is not available.
+            fn request_pages(&mut self, pages: usize) -> Option<usize> {
+                let prev_page = core::arch::wasm32::memory_grow(0, pages);
+                if prev_page == usize::MAX {
+                    return None;
+                }
+
+                prev_page.checked_mul(PAGE_SIZE)
+            }
+        } else if #[cfg(test)] {
+            /// Request a `pages` number of page sized sections of Wasm memory. Each page is `64KiB` in size.
+            ///
+            /// Returns `None` if a page is not available.
+            ///
+            /// This implementation is only meant to be used for testing, since we cannot (easily)
+            /// test the `wasm32` implementation.
+            fn request_pages(&mut self, _pages: usize) -> Option<usize> {
+                Some(self.upper_limit)
+            }
+        } else if #[cfg(all(not(test), feature = "std"))] {
```

Comment:
You do not need to check for `not(test)` here anymore because it is checked by the previous branch.